Add reusable block render tests

jaswrks opened the pull request:
## Description
Requested by @aduth in https://github.com/WordPress/gutenberg/pull/3882#issuecomment-350824562

## Types of changes
- Added tests for reusable block rendering.
  Includes tests for two special conditions validated by changes in https://github.com/WordPress/gutenberg/pull/3882

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

Commit:
Add reusable block render tests

noisysocks commented on phpunit/class-reusable-blocks-render-test.php:
```diff
@@ -0,0 +1,127 @@
+				),
+			),
+			'render_callback' => 'gutenberg_render_block_core_reusable_block',
+		) );
```
Could we avoid having to register this both in `blocks/library/block/index.php` and here in the test case?

Perhaps by using `WP_Block_Type_Registry::get_instance()->get_registered()` in the test cases instead of `$this->registry->get_instance()`?

jaswrks replied:
Thanks for the review!

Yes, that would have been my preference also, but there are [other existing tests affecting the main registry state](https://github.com/WordPress/gutenberg/blob/84487bab5258fe27897b601914fe5ee339bfcbe8/phpunit/class-dynamic-blocks-render-test.php#L41-L43), and I found that `core/block` was missing whenever this new test ran.

To prevent this from becoming an ongoing issue, I went ahead and straightened this out in the other tests, as an additional commit in this PR. See: https://github.com/WordPress/gutenberg/pull/4608/commits/6e8f9fb8e08f8eb5a70ada582c352f76a8671881

noisysocks reviewed (commented):
Excellent!

Commit:
Fix main block registry state in unit tests

Some unit tests were leaving the main registry in an
unexpected state, forcing other tests to avoid use
of the main registry altogether.

noisysocks commented on phpunit/class-reusable-blocks-render-test.php:
```diff
@@ -0,0 +1,121 @@
+		$this->registry = new WP_Block_Type_Registry();
+		$main_registry  = WP_Block_Type_Registry::get_instance();
+
+		$this->registry->register( 'core/block', $main_registry->get_registered( 'core/block' ) );
```
We can avoid having this proxy registry by using the main registry in the tests.

That is, we could delete this `setUp()` entirely and change our tests to look like this:

```php
public function test_render() {
    $block_type = WP_Block_Type_Registry::get_instance()->get_registered( 'core/block' );
    $output = $block_type->render( array( 'ref' => self::$block_id ) );
    $this->assertSame( '<p>Hello world!</p>', $output );
}
```

noisysocks commented on phpunit/class-registration-test.php:
```diff
@@ -13,31 +13,37 @@ class Registration_Test extends WP_UnitTestCase {
-			WP_Block_Type_Registry::get_instance()->unregister( $name );
+		$registry = WP_Block_Type_Registry::get_instance();
+
+		foreach ( array( 'core/dummy' ) as $name ) {
```
I think we may as well remove these `foreach` since we're only iterating over an array with one element.

noisysocks reviewed (changes requested):
This is looking good. Just a few minor things 🙂

Commit:
Streamline unit tests accessing block registry

- Remove unnecessary foreach loops.
- Instead of a new registry instance for tests,
  use main registry, since we only require read access.

jaswrks commented:
@noisysocks Thanks for the follow-up. Updated as requested.

Looks like some random Travis failures though. I'll rebase this to force a new CI build.

aduth commented on phpunit/class-prepare-for-js-test.php:
```diff
@@ -10,26 +10,15 @@
-			WP_Block_Type_Registry::get_instance()->unregister( $name );
-		}
+		$registry = WP_Block_Type_Registry::get_instance();
+		$registry->unregister( 'core/dummy' );
```
If we assume all tests should register a block only with this name, it might make for a slight improvement to move it to a static variable of the class, e.g. `protected static $block_name = 'core/dummy';`

aduth replied:
If we assume all tests should register a block only with this name, it might make for a slight improvement to move it to a static variable of the class, e.g. `protected static $block_name = 'core/dummy';`

aduth reviewed (approved):
Thanks! 👍

aduth commented on phpunit/class-dynamic-blocks-render-test.php:
```diff
@@ -38,9 +38,8 @@ function tearDown() {
 
 		$this->dummy_block_instance_number = 0;
 
-		foreach ( WP_Block_Type_Registry::get_instance()->get_all_registered() as $name => $block_type ) {
```
While not a behavior we should have wanted to depend on, noting that this change had an unintended consequence of having core server-registered blocks remain registered after the tests complete, which was not the case previously:

https://github.com/WordPress/gutenberg/blob/477e27ea1166bb5e856fa800bc639a25766146c6/phpunit/bootstrap.php#L31-L40

https://github.com/WordPress/gutenberg/blob/477e27ea1166bb5e856fa800bc639a25766146c6/lib/load.php#L24-L27

aduth commented on phpunit/class-reusable-blocks-render-test.php:
```diff
@@ -0,0 +1,97 @@
+/**
+ * Tests reusuable block rendering.
+ */
+class Reusuable_Blocks_Render_Test extends WP_UnitTestCase {
```
Typo: "Reusuable" -> "Reusable"

jaswrks replied:
Thank you. https://github.com/WordPress/gutenberg/pull/4736 corrects my mistake.
